aws/tailers/stack_event.go

```diff
@@ -0,0 +1,229 @@
+package awstailers
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"text/tabwriter"
+
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/fatih/color"
+	"github.com/wallix/awless/aws/services"
+)
+
+type stackEventTailer struct {
+	stackName        string
+	refresh          bool
+	watchDeployment  bool
+	refreshFrequency time.Duration
+	lastEventID      *string
+	nbEvents         int
+}
+
+type stackEvents []*cloudformation.StackEvent
+
+func coloredResourceStatus(str string) string {
+	switch {
+	case strings.HasSuffix(str, "_IN_PROGRESS"):
+		return color.New(color.FgYellow).SprintFunc()(str)
+	case strings.HasSuffix(str, "_COMPLETE"):
+		return color.New(color.FgGreen).SprintFunc()(str)
+	case strings.HasSuffix(str, "_FAILED"):
+		return color.New(color.FgRed).SprintFunc()(str)
+	default:
+		return str
+	}
+
+}
+
+func (e stackEvents) printReverse(w io.Writer) error {
+	tab := tabwriter.NewWriter(w, 35, 8, 0, '\t', 0)
+	for i := len(e) - 1; i >= 0; i-- {
+		_, err := fmt.Fprintf(tab, "%s\t%s\t%s\t%s\t", e[i].Timestamp.Format(time.RFC3339), *e[i].LogicalResourceId, *e[i].ResourceType, coloredResourceStatus(*e[i].ResourceStatus))
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(tab)
+	}
+	tab.Flush()
+	return nil
+}
+
+func NewCloudformationEventsTailer(stackName string, nbEvents int, enableRefresh bool, frequency time.Duration, watchDeployment bool) *stackEventTailer {
+	return &stackEventTailer{stackName: stackName, refresh: enableRefresh, refreshFrequency: frequency, nbEvents: nbEvents, watchDeployment: watchDeployment}
+}
+
+func (t *stackEventTailer) Name() string {
+	return "stack-events"
+}
+
+func (t *stackEventTailer) Tail(w io.Writer) error {
+	cfn, ok := awsservices.CloudformationService.(*awsservices.Cloudformation)
+	if !ok {
+		return fmt.Errorf("invalid cloud service, expected awsservices.Cloudformation, got %T", awsservices.CloudformationService)
+	}
+
+	if t.refreshFrequency < 5*time.Second {
+		return fmt.Errorf("invalid refresh frequency: %s", t.refreshFrequency)
+	}
+
+	if t.watchDeployment {
+		deploying, err := t.isStackBeingDeployed(cfn)
+		if err != nil {
+			return err
+		}
+
+		if !deploying {
+			return fmt.Errorf("Stack %s not being deployed at the moment", t.stackName)
+		}
+
+	} else {
+		if err := t.displayLastEvents(cfn, w); err != nil || !t.refresh {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(t.refreshFrequency)
+	defer ticker.Stop()
+	for range ticker.C {
+		if t.watchDeployment {
```

Comment:
Not fully seeing the value of the `watchDeployment`. Couldn't we always have the if clause (i.e. see if the deployment is finished) so that we can always exit and just print to Std.out to let know the user before exiting the CLI. The existing `--follow` on its own could then express the fact that we are watching the deployment

Reply:
Make sense.

How do you see `follow` behave if no deployment running? I would like to still throw an error since I don't see the point of following if nothing happens, this makes sense in terms of usage with the CI.

Reply:
Yes. Good idea. An explicit friendly error. You can use the exitOn(myerr) method for that.